Dev environment changes

hdonnay opened the pull request:
While investigating PROJQUAY-2152, I made a lot of changes.

This is both the jig I built to verify the notifier, and the changes to the dev environment to be less dependent on Docker specifically and utilize `docker-compose`'s native profile support.

hdonnay committed:
webhook: fix a code smell

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
webhook: add debug server

This script emulates a webhook receiver by retrieving every notification
and then deleting it.

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
httptransport: gofmt simplification

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
httptransport: error message correction

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
local-dev: clair: update default dev config

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
local-dev: traefik: use file configuration

This means the docker socket doesn't need to be mounted into the traefik
container, at the cost/advantage of putting the config in separate files
instead of container labels.

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
local-dev: grafana: update endpoint

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
local-dev: pgadmin: update for multiple databases

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
local-dev: prometheus: update endpoints

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
local-dev: quay: update for other config changes

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
docker-compose: use profiles and anchors

This makes the entire docker-compose file much easier to follow and
modify, IMO. It also has the notable benefit of `docker-compose up` just
working.

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
Makefile: update targets and docs

This removes the explicit `docker-compose` usage in favor of using the
profiles. It also changes the documentation target to be an actual
target.

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay commented:
TODO:

- [x] Verify the Quay changes (blocked on containers/podman#11966 being released)

hdonnay commented:
containers/podman#11966 is released in podman 3.4.1, so circling back to check on quay.

crozzy commented on Documentation/howto/testing.md:
````diff
@@ -29,72 +32,43 @@ See [Install Golang](https://golang.org/doc/install)
 After the local development environment successfully starts, the following infrastructure is available to you:
 
-```
-localhost:8080 --- Quay (single node, local storage)
````
Definitely an improvement here, although can we find somewhere for the credentials to avoid browsing / execing around for the defaults?

hdonnay committed:
docker-compose: add paused skopeo container

This is to work around Quay's inability to handle split-view DNS and
port-forwards. Using the API over localhost means clair can't talk to
quay, and using the container name means the host can't talk to quay
over the port forward. The solution I hit on is to just have a waiting
container with skopeo.

The documentation is updated to not recommend spinning up quay by
default, and to mention how to use this odd container arrangement.

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

crozzy commented on Makefile:
```diff
@@ -14,143 +14,43 @@
+	$(docker-compose) --profile $* up -d
+
+local-dev-quay: local-dev/clair/quay.yaml vendor
+	CLAIR_CONFIG=$(<F) $(docker-compose) --profile quay up -d
```
you think cleaning up the `local-dev/clair/quay.yaml` file here is a bit overreaching?

hdonnay replied:
What do you mean "cleaning up?" It needs to exist while things are running, and needs to be slightly different from the standard config. It being a proper target and dependency at least makes sure it gets updated as the standard config changes.

hdonnay committed:
docker-compose: add ActiveMQ container

This provides a container that speaks both STOMP and AMQP.

Signed-off-by: Hank Donnay <hdonnay@redhat.com>
